pkg/deploy/controller/deploymentconfig/controller.go

```diff
@@ -54,230 +56,193 @@ func (e transientError) Error() string {
+						fmt.Println("D")
+						activeReplicas = latestDesiredReplicas
+					}
 				}
 			}
 		}
-		allDetails = append(allDetails, details)
+	} else {
+		latestDesiredReplicas, latestHasDesiredReplicas := deployutil.DeploymentDesiredReplicas(latestDeployment)
+		if latestHasDesiredReplicas {
+			fmt.Println("E")
+			activeReplicas = latestDesiredReplicas
+		}
 	}
-
-	if len(allDetails) > 1 {
-		for i := range allDetails {
-			allDetails[i] = fmt.Sprintf("\t* %s", allDetails[i])
+	// Bring the config in sync with the deployment. Once we know the config
+	// accurately represents the desired replica count of the active deployment,
+	// we can safely reconcile deployments.
+	if config.Template.ControllerTemplate.Replicas != activeReplicas {
+		// TODO: Is this worth generating an event?
+		config.Template.ControllerTemplate.Replicas = activeReplicas
+		_, err := c.osClient.DeploymentConfigs(config.Namespace).Update(config)
+		if err != nil {
+			return err
 		}
-		// Prepend multiple errors warning
-		multipleErrWarning := fmt.Sprintf("Deployment config %q blocked by multiple errors:\n", config.Name)
-		allDetails = append([]string{multipleErrWarning}, allDetails...)
 	}
+	// Reconcile deployments. The active deployment follows the config, and all
+	// other deployments should be scaled to zero.
+	for _, deployment := range existingDeployments.Items {
+		isActiveDeployment := activeDeployment != nil && deployment.Name == activeDeployment.Name
 
-	return strings.Join(allDetails, "\n"), existingDeployments, latestDeploymentExists, nil
-}
-
-// updateDetails updates a deployment configuration with the provided details
-func (c *DeploymentConfigController) updateDetails(config *deployapi.DeploymentConfig, details string) (*deployapi.DeploymentConfig, error) {
-	if config.Details == nil {
-		config.Details = new(deployapi.DeploymentDetails)
-	}
-	if details != config.Details.Message {
-		config.Details.Message = details
-		return c.osClient.DeploymentConfigs(config.Namespace).Update(config)
+		oldReplicaCount := deployment.Spec.Replicas
+		newReplicaCount := 0
+		if isActiveDeployment {
+			newReplicaCount = activeReplicas
+		}
+		lastReplicas, hasLastReplicas := deployutil.DeploymentReplicas(&deployment)
+		// Only update if necessary.
+		if !hasLastReplicas || newReplicaCount != oldReplicaCount || lastReplicas != newReplicaCount {
+			deployment.Spec.Replicas = newReplicaCount
+			deployment.Annotations[deployapi.DeploymentReplicasAnnotation] = strconv.Itoa(newReplicaCount)
+			_, err := c.kubeClient.ReplicationControllers(deployment.Namespace).Update(&deployment)
+			if err != nil {
+				c.recorder.Eventf(config, "DeploymentScaleFailed",
+					"Failed to scale deployment %q from %d to %d: %s", deployment.Name, oldReplicaCount, newReplicaCount, err)
+				return err
+			}
+			// Only report scaling events if we changed the replica count.
+			// TODO: Is it worth making an event when we're just doing an annotation update?
```

Comment:
I'm leaning towards no, because the annotation is an semi-internal implementation detail. I suspect it's not really part of the life-cycle that a user or admin might care about.